File: cli/src/lib/collect/aggregators/tokens.ts
import { ModelUsageEntry, UsageBucket } from "../types.js";
import { calculateCost, COST_CURRENCY } from "../pricing.js";

const TOKEN_FIELDS = [
  "input_tokens",
  "output_tokens",
  "cache_read_input_tokens",
  "cache_creation_input_tokens",
] as const;

export function bucketKey(dims: {
  model: string;
  speed: string;
  service_tier: string;
  effort: string;
}): string {
  return `${dims.model}|speed=${dims.speed}|tier=${dims.service_tier}|effort=${dims.effort}`;
}

function emptyBucket(
  model: string,
  speed: string,
  service_tier: string,
  effort: string,
  currency: string
): UsageBucket {
  return {
    model,
    speed,
    service_tier,
    effort,
    api_calls: 0,
    input_tokens: 0,
    output_tokens: 0,
    cache_read_input_tokens: 0,
    cache_creation_input_tokens: 0,
    cost: 0,
    currency,
  };
}

function hasBillableTokens(usage: Record<string, unknown>): boolean {
  return TOKEN_FIELDS.some((field) => Number(usage[field] ?? 0) > 0);
}

function stableUsageJson(usage: Record<string, unknown>): string {
  const sorted: Record<string, unknown> = {};
  for (const key of Object.keys(usage).sort()) {
    sorted[key] = usage[key];
  }
  return JSON.stringify(sorted);
}

function unionAgents(a?: string[], b?: string[]): string[] | undefined {
  const all = [...(a ?? []), ...(b ?? [])];
  if (!all.length) return undefined;
  return [...new Set(all)].sort();
}

export function normalizeUsageBucketCurrency(bucket: UsageBucket): UsageBucket {
  return {
    ...bucket,
    currency: COST_CURRENCY,
  };
}

export function normalizeModelUsageCurrency(entry: ModelUsageEntry): ModelUsageEntry {
  return {
    ...entry,
    currency: COST_CURRENCY,
  };
}

/**
 * Processes Claude Code assistant events and aggregates usage into buckets.
 * Deduplicates by message.id if present, otherwise by (timestamp, usageJSON).
 * Returns a map from bucketKey to UsageBucket.
 */
export function aggregateUsageBuckets(
  events: Record<string, unknown>[],
  agent?: string,
  tokenSource?: string
): Record<string, UsageBucket> {
  const buckets: Record<string, UsageBucket> = {};
  const seen = new Set<string>();

  for (const event of events) {
    if (event["type"] !== "assistant") continue;

    const message = event["message"] as Record<string, unknown> | undefined;
    if (!message) continue;

    const usage = message["usage"] as Record<string, unknown> | undefined;
    if (!usage) continue;
    if (!hasBillableTokens(usage)) continue;

    const model = (message["model"] as string | undefined) ?? "";
    if (!model || model === "<synthetic>") continue;

    // Dedup key
    let dedupKey: string;
    const msgId = message["id"] as string | undefined;
    if (msgId) {
      dedupKey = msgId;
    } else {
      const ts = (event["timestamp"] as string | undefined) ?? "";
      dedupKey = `${ts}::${stableUsageJson(usage)}`;
    }

    if (seen.has(dedupKey)) continue;
    seen.add(dedupKey);

    // Extract dimension attributes
    const speed = (usage["speed"] as string | undefined) || "standard";
    const service_tier = (usage["service_tier"] as string | undefined) || "standard";
    const effort = (usage["effort"] as string | undefined) || (message["effort"] as string | undefined) || "default";

    const key = bucketKey({ model, speed, service_tier, effort });
    if (!buckets[key]) {
      buckets[key] = emptyBucket(model, speed, service_tier, effort, COST_CURRENCY);
      if (agent) buckets[key].agents = [agent];
      if (tokenSource) buckets[key].token_source = tokenSource;
    }

    const bucket = buckets[key];
    const input = (usage["input_tokens"] as number) ?? 0;
    const output = (usage["output_tokens"] as number) ?? 0;
    const cacheRead = (usage["cache_read_input_tokens"] as number) ?? 0;
    const cacheWrite = (usage["cache_creation_input_tokens"] as number) ?? 0;

    bucket.api_calls += 1;
    bucket.input_tokens += input;
    bucket.output_tokens += output;
    bucket.cache_read_input_tokens += cacheRead;
    bucket.cache_creation_input_tokens += cacheWrite;

    const callCost = calculateCost(model, { input_tokens: input, output_tokens: output, cache_read_input_tokens: cacheRead, cache_creation_input_tokens: cacheWrite }, { speed });
    if (callCost !== null) {
      bucket.cost += callCost;
    }
  }

  return buckets;
}

/**
 * Collapses per-dimension buckets into per-model totals.
 */
export function foldBucketsToModel(
  buckets: Record<string, UsageBucket>
): Record<string, ModelUsageEntry> {
  const models: Record<string, ModelUsageEntry> = {};

  for (const bucket of Object.values(buckets)) {
    const normalizedBucket = normalizeUsageBucketCurrency(bucket);
    const { model } = normalizedBucket;
    if (!models[model]) {
      models[model] = {
        api_calls: 0,
        input_tokens: 0,
        output_tokens: 0,
        cache_read_input_tokens: 0,
        cache_creation_input_tokens: 0,
        cost: 0,
        currency: COST_CURRENCY,
      };
    }
    const entry = models[model];
    entry.api_calls += normalizedBucket.api_calls;
    entry.input_tokens += normalizedBucket.input_tokens;
    entry.output_tokens += normalizedBucket.output_tokens;
    entry.cache_read_input_tokens += normalizedBucket.cache_read_input_tokens;
    entry.cache_creation_input_tokens += normalizedBucket.cache_creation_input_tokens;

    entry.cost += normalizedBucket.cost;
    entry.agents = unionAgents(entry.agents, normalizedBucket.agents);
  }

  return models;
}

/**
 * Merge two bucket maps — sum all numeric fields, sum pre-calculated costs.
 */
export function mergeUsageBuckets(
  a: Record<string, UsageBucket>,
  b: Record<string, UsageBucket>
): Record<string, UsageBucket> {
  const result: Record<string, UsageBucket> = {};

  for (const [key, bucket] of Object.entries(a)) {
    result[key] = normalizeUsageBucketCurrency(bucket);
  }

  for (const [key, bucket] of Object.entries(b)) {
    const normalizedBucket = normalizeUsageBucketCurrency(bucket);
    if (!result[key]) {
      result[key] = normalizedBucket;
    } else {
      const existing = result[key];
      existing.api_calls += normalizedBucket.api_calls;
      existing.input_tokens += normalizedBucket.input_tokens;
      existing.output_tokens += normalizedBucket.output_tokens;
      existing.cache_read_input_tokens += normalizedBucket.cache_read_input_tokens;
      existing.cache_creation_input_tokens += normalizedBucket.cache_creation_input_tokens;

      existing.cost += normalizedBucket.cost;
      existing.agents = unionAgents(existing.agents, normalizedBucket.agents);
    }
  }

  return result;
}

/**
 * Sum total cost by currency across all buckets.
 */
export function sumCostByCurrency(
  buckets: Record<string, UsageBucket>
): Record<string, number> {
  const totals: Record<string, number> = {};

  for (const bucket of Object.values(buckets)) {
    const normalizedBucket = normalizeUsageBucketCurrency(bucket);
    totals[COST_CURRENCY] = (totals[COST_CURRENCY] ?? 0) + normalizedBucket.cost;
  }

  return totals;
}

File: cli/python/ai_coding_reports/aggregators/tokens.py
"""Token usage aggregation with deduplication and model/speed/tier buckets."""

from __future__ import annotations

import json
from collections import defaultdict

from ai_coding_reports.utils.pricing import (
    calculate_cost_for_model,
    get_currency_for_model,
)

_SKIP_MODELS = frozenset({"<synthetic>", ""})

_TOKEN_FIELDS = (
    "input_tokens",
    "output_tokens",
    "cache_read_input_tokens",
    "cache_creation_input_tokens",
)


def _has_billable_tokens(usage: dict) -> bool:
    return any(usage.get(k, 0) for k in _TOKEN_FIELDS)


def _dedup_key(evt: dict, msg: dict, usage: dict) -> tuple:
    mid = msg.get("id")
    if mid:
        return ("id", mid)
    return ("fp", evt.get("timestamp", ""), json.dumps(usage, sort_keys=True))


def usage_dimensions(msg: dict, usage: dict) -> dict[str, str]:
    """Extract billing dimensions from an assistant message."""
    return {
        "model": msg.get("model", "unknown"),
        "speed": usage.get("speed") or "standard",
        "service_tier": usage.get("service_tier") or "standard",
        "effort": usage.get("effort") or msg.get("effort") or "default",
    }


def bucket_key(dims: dict[str, str]) -> str:
    return (
        f"{dims['model']}|speed={dims['speed']}|tier={dims['service_tier']}"
        f"|effort={dims['effort']}"
    )


def _empty_bucket() -> dict:
    return {
        "api_calls": 0,
        "input_tokens": 0,
        "output_tokens": 0,
        "cache_read_input_tokens": 0,
        "cache_creation_input_tokens": 0,
    }


def aggregate_usage_buckets(events: list[dict]) -> dict[str, dict]:
    """Aggregate assistant usage by model × speed × tier × effort.

    Returns {bucket_key: {model, speed, service_tier, effort, tokens, cost, ...}}.
    """
    buckets: dict[str, dict] = defaultdict(_empty_bucket)
    dims_by_key: dict[str, dict[str, str]] = {}
    seen: set[tuple] = set()

    for evt in events:
        if evt.get("type") != "assistant":
            continue
        msg = evt.get("message", {})
        if isinstance(msg, str):
            continue

        dims = usage_dimensions(msg, msg.get("usage") or {})
        if dims["model"] in _SKIP_MODELS:
            continue

        usage = msg.get("usage", {})
        if not usage or not _has_billable_tokens(usage):
            continue

        key = _dedup_key(evt, msg, usage)
        if key in seen:
            continue
        seen.add(key)

        bkey = bucket_key(dims)
        dims_by_key[bkey] = dims
        entry = buckets[bkey]
        entry["api_calls"] += 1
        entry["input_tokens"] += usage.get("input_tokens", 0)
        entry["output_tokens"] += usage.get("output_tokens", 0)
        entry["cache_read_input_tokens"] += usage.get("cache_read_input_tokens", 0)
        entry["cache_creation_input_tokens"] += usage.get(
            "cache_creation_input_tokens", 0
        )

    result: dict[str, dict] = {}
    for bkey, entry in dict(buckets).items():
        dims = dims_by_key[bkey]
        model = dims["model"]
        speed = dims["speed"]
        cost = calculate_cost_for_model(model, entry, speed=speed)
        currency = get_currency_for_model(model, speed=speed)
        result[bkey] = {
            **dims,
            **entry,
            "cost": cost if cost is not None else "unknown",
            "currency": currency,
        }

    return result


def buckets_from_breakdown(rows: list[dict]) -> dict[str, dict]:
    """Convert session usage_breakdown list to keyed bucket dict."""
    out: dict[str, dict] = {}
    for row in rows:
        dims = {
            "model": row.get("model", "unknown"),
            "speed": row.get("speed", "standard"),
            "service_tier": row.get("service_tier", "standard"),
            "effort": row.get("effort", "default"),
        }
        out[bucket_key(dims)] = row
    return out


def fold_buckets_to_model(buckets: dict[str, dict]) -> dict[str, dict]:
    """Merge dimension buckets into per-model totals (for daily.json compat).

    Token counts are summed per model. Cost is the sum of each dimension
    bucket's pre-calculated cost (preserves speed/tier/effort pricing).
    """
    by_model: dict[str, dict] = defaultdict(_empty_bucket)
    cost_by_model_currency: dict[tuple[str, str], float] = defaultdict(float)
    currencies: dict[str, str] = {}

    for entry in buckets.values():
        model = entry.get("model", "unknown")
        m = by_model[model]
        m["api_calls"] += entry.get("api_calls", 0)
        m["input_tokens"] += entry.get("input_tokens", 0)
        m["output_tokens"] += entry.get("output_tokens", 0)
        m["cache_read_input_tokens"] += entry.get("cache_read_input_tokens", 0)
        m["cache_creation_input_tokens"] += entry.get(
            "cache_creation_input_tokens", 0
        )
        currency = entry.get("currency", "?")
        currencies[model] = currency
        cost = entry.get("cost")
        if isinstance(cost, (int, float)):
            cost_by_model_currency[(model, currency)] += cost

    result: dict[str, dict] = {}
    for model, entry in dict(by_model).items():
        currency = currencies.get(model, "?")
        model_cost = cost_by_model_currency.get((model, currency), 0.0)
        has_bucket_cost = any(
            isinstance(entry.get("cost"), (int, float))
            for entry in buckets.values()
            if entry.get("model") == model
        )
        if has_bucket_cost:
            cost: float | str = round(model_cost, 2)
        else:
            cost = "unknown"
        result[model] = {
            **entry,
            "cost": cost,
            "currency": currency,
        }
    return result


def aggregate_tokens(events: list[dict]) -> dict[str, dict]:
    """Backward-compatible alias: per-model totals (folded from dimension buckets)."""
    return fold_buckets_to_model(aggregate_usage_buckets(events))


def merge_priced_buckets(a: dict[str, dict], b: dict[str, dict]) -> dict[str, dict]:
    """Merge bucket dicts by summing tokens and pre-calculated costs (no repricing).

    Use when merging session ``usage_breakdown`` rows that already have
    speed/tier/effort-aware costs.
    """
    merged: dict[str, dict] = {k: dict(v) for k, v in a.items()}
    for bkey, eb in b.items():
        if bkey not in merged:
            merged[bkey] = dict(eb)
            continue
        ea = merged[bkey]
        for field in _TOKEN_FIELDS:
            ea[field] = ea.get(field, 0) + eb.get(field, 0)
        ea["api_calls"] = ea.get("api_calls", 0) + eb.get("api_calls", 0)
        cost_a, cost_b = ea.get("cost"), eb.get("cost")
        if isinstance(cost_a, (int, float)) and isinstance(cost_b, (int, float)):
            ea["cost"] = round(cost_a + cost_b, 4)
        elif isinstance(cost_b, (int, float)):
            ea["cost"] = cost_b
        elif isinstance(cost_a, (int, float)):
            ea["cost"] = cost_a
    return merged


def merge_buckets(a: dict[str, dict], b: dict[str, dict]) -> dict[str, dict]:
    """Merge two bucket dicts (sum tokens; recompute cost per bucket)."""
    merged: dict[str, dict] = {}
    all_keys = set(a) | set(b)

    for bkey in all_keys:
        ea, eb = a.get(bkey), b.get(bkey)
        if ea and eb:
            base = {k: ea.get(k, 0) + eb.get(k, 0) for k in _TOKEN_FIELDS}
            base["api_calls"] = ea.get("api_calls", 0) + eb.get("api_calls", 0)
            dims = {k: ea.get(k) for k in ("model", "speed", "service_tier", "effort")}
        elif ea:
            base = {k: ea.get(k, 0) for k in (*_TOKEN_FIELDS, "api_calls")}
            dims = {k: ea.get(k) for k in ("model", "speed", "service_tier", "effort")}
        else:
            base = {k: eb.get(k, 0) for k in (*_TOKEN_FIELDS, "api_calls")}
            dims = {k: eb.get(k) for k in ("model", "speed", "service_tier", "effort")}

        model = dims.get("model", "unknown")
        speed = dims.get("speed", "standard")
        cost = calculate_cost_for_model(model, base, speed=speed)
        currency = get_currency_for_model(model, speed=speed)
        merged[bkey] = {
            **dims,
            **base,
            "cost": cost if cost is not None else "unknown",
            "currency": currency,
        }

    return merged


def sum_cost_by_currency(buckets: dict[str, dict]) -> dict[str, float]:
    """Sum numeric costs grouped by currency symbol."""
    totals: dict[str, float] = defaultdict(float)
    for entry in buckets.values():
        cost = entry.get("cost")
        if not isinstance(cost, (int, float)):
            continue
        currency = entry.get("currency", "?")
        totals[currency] += cost
    return {c: round(v, 4) for c, v in totals.items()}


def sum_tokens(model_usage: dict) -> dict[str, int]:
    """Sum input/output tokens across all models or buckets."""
    totals = {"input_tokens": 0, "output_tokens": 0, "cache_read": 0, "cache_write": 0}
    for entry in model_usage.values():
        totals["input_tokens"] += entry.get("input_tokens", 0)
        totals["output_tokens"] += entry.get("output_tokens", 0)
        totals["cache_read"] += entry.get("cache_read_input_tokens", 0)
        totals["cache_write"] += entry.get("cache_creation_input_tokens", 0)
    return totals
